File: problems/intersectionOfTwoArrays/c++/intersection.cpp
/* Usage with CLI on Mac
```bash
g++ -std=c++11 -o intersection intersection.cpp
./intersection
```
*/
#include<iostream>
#include<string>
#include<vector>
#include<unordered_set>

using namespace std;

class Solution {
public:
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        unordered_set<int> unique1(nums1.begin(), nums1.end());
        vector<int> result;
        for (auto num : nums2)
            if (unique1.erase(num) > 0)
                result.push_back(num);
        return result;
    }
};

int main() {
    vector<int> num1 = {1, 2, 3};
    vector<int> num2 = {2, 3, 4};
    Solution obj;

    vector<int> result = obj.intersection(num1, num2);

    cout << "The answer is: ";
    for (int num : result) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: problems/intersectionOfTwoArrays/python/intersection.py
from typing import List


class Solution1:
    """
    Complexity
    ----------
    - TC: O(n+m)
    - SC: O(n+m)
    """

    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        unique1 = set(nums1)
        unique2 = set(nums2)

        if unique1 > unique2:
            unique1, unique2 = unique2, unique1
        return [num for num in unique1 if num in unique2]


class Solution2:
    """
    Complexity
    ----------
    - TC: O(n+m)
    - SC: O(n+m)
    """

    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        return list(set(nums1) & set(nums2))


class Solution3:
    """
    Complexity
    ----------
    - TC: O(n+m)
    - SC: O(n+m)
    """

    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        return list(set(nums1).intersection(nums2))


if __name__ == "__main__":
    num1 = [4, 5, 6, 7]
    num2 = [7, 7]

    print(Solution1().intersection(num1, num2))
    print(Solution2().intersection(num1, num2))
    print(Solution3().intersection(num1, num2))
